core/src/main/scala/schema/provider/MetadataSchemaProvider.scala

```diff
@@ -52,14 +58,15 @@ class MetadataSchemaProvider[M[_]](cluster: Cluster)(implicit H: FSHandler[Clust
         M.flatMap(finalizer)(_ => e.fold(M.raiseError, M.pure))
       }
 
-    def metadataF[F[_]](implicit clusterAPI: ClusterAPI[F]): FreeS[F, Metadata] =
-      clusterAPI.connect *> clusterAPI.metadata
+    def metadataF: FreeS[ClusterAPI.Op, Metadata] = API.connect *> API.metadata
+
+    def closeF: FreeS[ClusterAPI.Op, Unit] = API.close
 
-    def closeF[F[_]](implicit clusterAPI: ClusterAPI[F]): FreeS[F, Unit] =
-      clusterAPI.close
+    M.flatMap(clusterProvider) { cluster =>
+      implicit val H: FSHandler[ClusterAPI.Op, M] =
+        clusterAPIHandler[M] andThen apiInterpreter[M, Cluster](cluster)
 
-    M.flatMap(guarantee(metadataF[ClusterAPI.Op].interpret[M], closeF[ClusterAPI.Op].interpret[M])) {
-      metadata =>
+      M.flatMap(guarantee(metadataF.interpret[M], closeF.interpret[M])) { metadata =>
```

Comment:
Here we have `metadataF` and `closeF` with two different interpretations. Can we compose both and interpret just once?

Reply:
They are two different operations for the `guarantee` method, but I'll take a look